File: array/186_Reverse_Words_in_a_String_II.c
void reverse(char *s, int start, int end){
    while(start<end){
        char tmp = s[start];
        s[start] = s[end];
        s[end] =tmp;
        start ++;
        end--;
    }
    return;
}

void reverseWords(char *s) {
    if(!s || strlen(s)==0) return s;
    int len = strlen(s);
    reverse(s, 0, len-1);
    int start =0;
    int end = start;
    while(end<len){
        if(s[end] == ' '){
            reverse(s, start, end-1);
            start = end+1;
        }
        end ++;
    }
    reverse(s, start, len - 1);//reverse the last word, if there is only one word this will solve the corner case
    return s;
}

File: java/186. Reverse Words in a String II.java
class Solution {
    public void reverseWords(char[] str) {
        // reverse the words and then reverse the whole string
        int size  = str.length;
        int start  = 0;

        for (int i = 0; i <= size ; i++) {
            if (i == size  || str[i] == ' ') {
                inPlaceReverse(start, i-1, str);
                start = i+1;
            } 
        }
        inPlaceReverse(0, size - 1, str);

    }
    
    private void inPlaceReverse(int start, int end, char[] str) {
        while (start < end) {
            char tmp = str[start];
            str[start] = str[end];
            str[end] =  tmp;
            end --;
            start ++;
          }
    }
}
